pkg/util/shell/shell_test.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright (C) 2017 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shell
+
+import (
+	"os"
+	"testing"
+)
+
+type shellTestCase struct {
+	name         string
+	cmdLine      string
+	expectedHint string
+}
+
+type prefixSuffixDelimiterTestCase struct {
+	shellName string
+	prefix    string
+	suffix    string
+	pathVar   bool
+	delimiter string
+}
+
+func TestUnknownShell(t *testing.T) {
+	expectedShellName := "foo"
+	defer func(shell string) { os.Setenv("SHELL", shell) }(os.Getenv("SHELL"))
+	os.Setenv("SHELL", "/bin/"+expectedShellName)
+
+	shell, err := GetShell("")
+	if err != nil {
+		t.Fatal("Unexpected err : ", err.Error())
+	}
+
+	if shell != expectedShellName {
+		t.Fatalf("Expected shell to be %s but %s", expectedShellName, shell)
+	}
+}
+
+func TestKnownShell(t *testing.T) {
+	expectedShellName := "bash"
+	shell, err := GetShell("bash")
+	if err != nil {
+		t.Fatal("Unexpected err : ", err.Error())
+	}
+
+	if shell != expectedShellName {
+		t.Fatalf("Expected shell to be %s but %s", expectedShellName, shell)
+	}
+}
+
+func TestNoProxyFromEnvWhenUsedLowerCase(t *testing.T) {
+	expectedNoProxyVarName := "no_proxy"
+	expectedNoProxyVarValue := "FOO_BAR"
+	defer func(val string) { os.Setenv(expectedNoProxyVarName, val) }(os.Getenv(expectedNoProxyVarName))
+	os.Setenv(expectedNoProxyVarName, expectedNoProxyVarValue)
+
+	_, noProxyValue := FindNoProxyFromEnv()
+	if noProxyValue != expectedNoProxyVarValue {
+		t.Fatalf("Expected no proxy var value as %s but got %s", expectedNoProxyVarValue, noProxyValue)
+	}
+}
+
+func TestNoProxyFromEnvWhenUsedUpperCase(t *testing.T) {
+	expectedNoProxyVarName := "NO_PROXY"
+	expectedNoProxyVarValue := "FOO_BAR"
+	defer func(val string) { os.Setenv(expectedNoProxyVarName, val) }(os.Getenv(expectedNoProxyVarName))
+	os.Setenv(expectedNoProxyVarName, expectedNoProxyVarValue)
+
+	_, noProxyValue := FindNoProxyFromEnv()
+	if noProxyValue != expectedNoProxyVarValue {
+		t.Fatalf("Expected no proxy var value as %s but got %s", expectedNoProxyVarValue, noProxyValue)
+	}
+}
+
+func TestGenerateUsageHint(t *testing.T) {
```

Comment:
I was hoping to test the actual final output of _oc-env_ and _docker-env_. This should not be so hard really. Rename _executeOcTemplateStdout_ to _executeOcTemplate_ and pass the writer into the function. For production code it is _os.Stdout_. For test purposes you can pass a string buffer. This way you can test the final output of the commands.

You should be able to do something like:

     shellCfg, err = getOcShellConfig("/home/john/.minishift/oc/1.5.1/oc", "fish")
     executeOcTemplateStdout(shellCfg, buffer)
    // make assertions against the buffer

I think such a test would be more valuable. It is almost like an integration test, except that we can skip the running VM.

WDYT?